### PR TITLE
feat(security): Deploy new security-bootstrapper service

### PR DESCRIPTION
**Pull request by jim-wang-intel**
Docker-compose deploys with a new security-bootstrapper service which controls the security bootstrapping steps for various phases.
The details are summarized in [ADR secure bootstrapping](https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/adr/security/0009-Secure-Bootstrapping.md).

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/developer-scripts/blob/master/.github/Contributing.md.

## What is the current behavior?

This is no security-bootstrapping service currently. 

## What is the new behavior?
New security-bootstrapper service with service-gating mechanism in installation phase.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?
_In draft mode: this won't function properly until [edgex-go's implementation PR](https://github.com/edgexfoundry/edgex-go/pull/2970) merged into branch master_

## Other information
Pending until this PR resolving the ARM64 type of tests work.  https://github.com/edgexfoundry/edgex-go/pull/3082

### CODE REVIEW

**Review by bnevis-i — COMMENTED** [review]
LGTM

File: compose-builder/add-security.yml [hunk 1]
```diff
@@ -257,62 +218,72 @@ services:
       - common-security.env
     environment:
       SECRETSTORE_TOKENFILE: /tmp/edgex/secrets/edgex-support-notifications/secrets-token.json
+    entrypoint: ["/edgex-init/ready_to_run_wait_install.sh"]
+    command: "/support-notifications ${DEFAULT_EDGEX_RUN_CMD_PARMS}"
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:222]: Perfect!

**Comment by lenny-intel** [comment]
@jim-wang-intel , I merge my PR. Please rebase and address any changes needed in the new partial compose files for services that are expected to run in secure mode.

File: compose-builder/add-security.yml [hunk 2]
```diff
@@ -20,36 +20,55 @@ version: '3.7'
...
   postgres-data:
 
 services:
   security-bootstrapper:
-    image: gcr.io/google_containers/defaultbackend:1.4
-    container_name: edgex-security-installer
-    hostname: edgex-security-installer
+    image: ${CORE_EDGEX_REPOSITORY}/docker-security-bootstrapper-go${ARCH}:${CORE_EDGEX_VERSION}${DEV}
+    container_name: edgex-security-stage-gate
+    hostname: edgex-security-stage-gate
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:35]: IMO, these should be `edgex-security-bootstrapper`

File: compose-builder/add-security.yml [hunk 3]
```diff
@@ -97,49 +106,24 @@ services:
     image: ${CORE_EDGEX_REPOSITORY}/docker-security-secretstore-setup-go${ARCH}:${CORE_EDGEX_VERSION}${DEV}
     container_name: edgex-vault-worker
     hostname: edgex-vault-worker
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:108]: IMO, we should change these to `edgex-secretstore-setup` and the service name above to `secretstore-setup` so names are consistent.

File: compose-builder/add-security.yml [hunk 4]
```diff
@@ -225,94 +201,106 @@ services:
     image: ${CORE_EDGEX_REPOSITORY}/docker-security-proxy-setup-go${ARCH}:${CORE_EDGEX_VERSION}${DEV}
     container_name: edgex-proxy
     hostname: edgex-proxy
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:203]: Another place we need to fix naming for consistency. These should be `edgex-proxy-setup` and service should be `proxy-setup`

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: compose-builder/add-security.yml (hunk 3 above)
Comment (jim-wang-intel) [reply compose-builder/add-security.yml]: changed

File: compose-builder/add-security.yml (hunk 4 above)
Comment (jim-wang-intel) [reply compose-builder/add-security.yml]: fixed

File: compose-builder/README.md [hunk 5]
```diff
@@ -86,6 +86,8 @@ This folder contains the following environment files:
     This file contains the common environment overrides used by all Edgex services.
 - **common-security.env**
     This file contains the common security related environment overrides used by many Edgex services.
+- **common-sec-bootstrapper.env**
```
Comment (lenny-intel) [inline compose-builder/README.md:89]: IMO, we show call this common-sec-stage-gate.env which matches your description

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Looks good, just one minor name suggestion.

File: compose-builder/README.md (hunk 5 above)
Comment (jim-wang-intel) [reply compose-builder/README.md:89]: ok, I'll rename it and rebase push once beau's PR is in.

Comment (jim-wang-intel) [reply compose-builder/README.md]: renamed

**Review by lenny-intel — COMMENTED** [review]
LGTM, but want to test in my environment first.

**Comment by jim-wang-intel** [comment]
> LGTM, but want to test in my environment first.

ok

**Comment by cherrycl** [comment]
@lenny-intel @jim-wang-intel The smoke test fails because the name changed on both services,  vault-worker and edgex-proxy. Do they have to do name changing? If yes, I can open a PR on edgex-taf repo to modify them.

**Comment by lenny-intel** [comment]
> If yes, I can open a PR on edgex-taf repo to modify them.

Yes, please do. we are making the naming of the service more consistent. We just merged the related edgex-go PR and will be looking to merge this one once it passes smock tests. THX!

**Comment by cherrycl** [comment]
@jim-wang-intel I found the below log on `edgex-security-bootstrapper` when running taf test. It seems to try to connect `consul`. Does it need to set depends-on with `consul`?

`level=INFO ts=2021-01-22T09:40:23.436782771Z app=edgex-security-bootstrapper source=client.go:50 msg="TCP server edgex-core-consul:54324 may be not ready yet, retry in 1 second"`

**Comment by cherrycl** [comment]
> @jim-wang-intel I found the below log on `edgex-security-bootstrapper` when running taf test. It seems to try to connect `consul`. Does it need to set depends-on with `consul`?
> 
> `level=INFO ts=2021-01-22T09:40:23.436782771Z app=edgex-security-bootstrapper source=client.go:50 msg="TCP server edgex-core-consul:54324 may be not ready yet, retry in 1 second"`

When running TAF test, I found the `edgex-proxy-setup` cannot connect `edgex-security-bootstrapper` on time, because the `edgex-security-bootstrapper` start-up completed time later than `edgex-proxy-setup` timeout.
That also causes the TAF script failure.

**Comment by jim-wang-intel** [comment]
> > @jim-wang-intel I found the below log on `edgex-security-bootstrapper` when running taf test. It seems to try to connect `consul`. Does it need to set depends-on with `consul`?
> > `level=INFO ts=2021-01-22T09:40:23.436782771Z app=edgex-security-bootstrapper source=client.go:50 msg="TCP server edgex-core-consul:54324 may be not ready yet, retry in 1 second"`
> 


@cherrycl That is expected. The essence of `security-bootstrapper` is to gate the right startup sequence for services.  So in this particular case, it awaits the `Consul` service to be ready. 


> When running TAF test, I found the `edgex-proxy-setup` cannot connect `edgex-security-bootstrapper` on time, because the `edgex-security-bootstrapper` start-up completed time later than `edgex-proxy-setup` timeout.
> That also causes the TAF script failure.

@cherrycl  not sure what this test case is. the `edgex-proxy-setup` is gated until `edgex-scurity-bootstrapper` reaches to the ready-to-run state. To reach to the `ready-to-run` state, all security related services have to be started up and running.  These include `vault`, `secretstore-setup`,  `consul`, `redis` database, and `kong-db`.   If time is not enough for gating these services in TAF tests, we can change the value of env var. `SECTY_BOOTSTRAP_GATING_TIMEOUT_DURATION` in `common-sec-stage-gate.env` file, in which it is set to `60s` right now.

**Comment by lenny-intel** [comment]
@cherrycl , I think the root cause may be that the TAF scripts are using `docker run` to start each service individually, rather that use `docker-compose -p edgex -f <compose-file> up` as the compose files are designed for. This may be causing the delays and failure you are seeing and also requires the TAF scripts to explicitly know the `service` names, causing maintenance when changed as we are seeing now.
https://github.com/edgexfoundry/edgex-taf/blob/master/TAF/utils/scripts/docker/deploy-edgex.sh 
https://github.com/edgexfoundry/edgex-taf/blob/master/TAF/utils/scripts/docker/deploy-services.sh
https://github.com/edgexfoundry/edgex-taf/blob/master/TAF/utils/scripts/docker/deploy-device-service.sh

Can't the TAF scripts just depend on `docker-compose -p edgex -f <compose-file> up` to start all the required services based on the appropriate compose file that was downloaded?

**Comment by cherrycl** [comment]
@lenny-intel May we keep the current deploy steps? Recently security test got failure because of the following vault-worker error. We found this PR seems to fix the error. Is this PR ready to merge?
`/usr/local/bin/entrypoint.sh: line 39: /edgex-init/security-bootstrapper: not found`

I will investigate the starting all services at once on TAF later.

**Comment by lenny-intel** [comment]
recheck

**Comment by lenny-intel** [comment]
> We found this PR seems to fix the error. Is this PR ready to merge?

Yes, once it passes smoke tests.

> I will investigate the starting all services at once on TAF later.

Yes, that is fine.

**Comment by lenny-intel** [comment]
recheck

**Comment by jim-wang-intel** [comment]
Hi @cherrycl  the taf tests still failed and probably due to the V2 DTO event stuff recently changed. Please update that and hopefully taf tests can all passed.

**Comment by cherrycl** [comment]
@jim-wang-intel Ginny is helping with app-service failure. 
But the arm security test also failed. The `vault` failed to start and got the following log.
```
Script for waiting security bootstrapping on Vault
Tue Jan 26 03:15:50 UTC 2021 VAULT_LOCAL_CONFIG: 
listener "tcp" { 
               address = "edgex-vault:8200" 
               tls_disable = "1" 
               cluster_address = "edgex-vault:8201" 
           } 
          backend "file" {
               path = "/vault/file"
           } 
           default_lease_ttl = "168h" 
           max_lease_ttl = "720h"
 
Tue Jan 26 03:15:50 UTC 2021 Executing dockerize on vault server with waiting on     tcp://edgex-security-bootstrapper:54321
/edgex-init/dockerize: line 1: syntax error: unexpected word (expecting ")")
```
Do you have any idea for the error?

**Comment by cherrycl** [comment]
@jim-wang-intel The failed test for app-service is fixed by [edgex-taf PR #282](https://github.com/edgexfoundry/edgex-taf/pull/282).

**Comment by lenny-intel** [comment]
recheck

**Comment by jim-wang-intel** [comment]
> @jim-wang-intel Ginny is helping with app-service failure.
> But the arm security test also failed. The `vault` failed to start and got the following log.
> 
> ```
> Script for waiting security bootstrapping on Vault
> Tue Jan 26 03:15:50 UTC 2021 VAULT_LOCAL_CONFIG: 
> listener "tcp" { 
>                address = "edgex-vault:8200" 
>                tls_disable = "1" 
>                cluster_address = "edgex-vault:8201" 
>            } 
>           backend "file" {
>                path = "/vault/file"
>            } 
>            default_lease_ttl = "168h" 
>            max_lease_ttl = "720h"
>  
> Tue Jan 26 03:15:50 UTC 2021 Executing dockerize on vault server with waiting on     tcp://edgex-security-bootstrapper:54321
> /edgex-init/dockerize: line 1: syntax error: unexpected word (expecting ")")
> ```
> 
> Do you have any idea for the error?



> @jim-wang-intel Ginny is helping with app-service failure.
> But the arm security test also failed. The `vault` failed to start and got the following log.
> 
> ```
> Script for waiting security bootstrapping on Vault
> Tue Jan 26 03:15:50 UTC 2021 VAULT_LOCAL_CONFIG: 
> listener "tcp" { 
>                address = "edgex-vault:8200" 
>                tls_disable = "1" 
>                cluster_address = "edgex-vault:8201" 
>            } 
>           backend "file" {
>                path = "/vault/file"
>            } 
>            default_lease_ttl = "168h" 
>            max_lease_ttl = "720h"
>  
> Tue Jan 26 03:15:50 UTC 2021 Executing dockerize on vault server with waiting on     tcp://edgex-security-bootstrapper:54321
> /edgex-init/dockerize: line 1: syntax error: unexpected word (expecting ")")
> ```
> 
> Do you have any idea for the error?

Hi @cherrycl  yea, this is a defect  in the current dockerfile of `security-bootstrapper` only supporting amd64 arch.  We will add ARM64 support into Dockerfile of that. Thanks!

**Comment by lenny-intel** [comment]
recheck

**Comment by lenny-intel** [comment]
recheck

**Comment by cherrycl** [comment]
@jim-wang-intel Thanks. The vault error was not found now, but kong appears a similar error and the startup failure.
```
Script for waiting security bootstrapping on Kong
Wed Jan 27 01:04:06 UTC 2021 Executing dockerize with waiting on tcp://edgex-security-bootstrapper:54329
/edgex-init/kong_wait_install.sh: line 31: /edgex-init/dockerize: No such file or directory
```

**Comment by jim-wang-intel** [comment]
> @jim-wang-intel Thanks. The vault error was not found now, but kong appears a similar error and start fail.
> 
> ```
> Script for waiting security bootstrapping on Kong
> Wed Jan 27 01:04:06 UTC 2021 Executing dockerize with waiting on tcp://edgex-security-bootstrapper:54329
> /edgex-init/kong_wait_install.sh: line 31: /edgex-init/dockerize: No such file or directory
> ```

@cherrycl  This is a different type of error and very weird error. `/edgex-init/dockerize`  should be in there as volume mount is shared.  not sure why this is happening.

**Comment by jim-wang-intel** [comment]
> > @jim-wang-intel Thanks. The vault error was not found now, but kong appears a similar error and start fail.
> > ```
> > Script for waiting security bootstrapping on Kong
> > Wed Jan 27 01:04:06 UTC 2021 Executing dockerize with waiting on tcp://edgex-security-bootstrapper:54329
> > /edgex-init/kong_wait_install.sh: line 31: /edgex-init/dockerize: No such file or directory
> > ```
> 
> @cherrycl This is a different type of error and very weird error. `/edgex-init/dockerize` should be in there as volume mount is shared. not sure why this is happening.

@cherrycl  maybe the way TAF tests using `docker run` has created some timing issue among different containers regarding to the docker volume sharing among them.  I think the better way to resolve this may be like @lenny-intel 's suggestion to use `docker-compose up` to run instead of individual `docker run`.  The original way of `docker run` was in place due to the startup sequence was not guarded against several containers in the past but now the `security-bootstrapper` is making that startup sequence right.

**Comment by jinlinGuan** [comment]
@jim-wang-intel I got the same error when running `docker-compose -f docker-compose-nexus-arm64.yml up -d`.

**Comment by jim-wang-intel** [comment]
recheck

**Comment by jim-wang-intel** [comment]
rebased and squahsed. ready to merged once build done.

**Review by bnevis-i — APPROVED** [review]

**Review by lenny-intel — APPROVED** [review]

**Pull request merged**